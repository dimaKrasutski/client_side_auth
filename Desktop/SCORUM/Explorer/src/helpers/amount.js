import { compose, curry } from 'ramda';
import { MAX_FRACTION_DIGITS, DEFAULT_LOCALE, MIN_FRACTION_DIGITS } from '../constants';

export const processAmountValue = amount => amount.replace(/[^\d.]/g, '');
export const amountStrToNum = compose(Number.parseFloat, processAmountValue);
export const scrToFixed = amount => `${Number.parseFloat(amount).toFixed(MAX_FRACTION_DIGITS)} SCR`;

/* eslint-disable prefer-template */
export const floorFloatAmount = (val, decimals) =>
  Number(Math.floor(Number.parseFloat(val).toFixed(MAX_FRACTION_DIGITS) + 'e' + decimals) + 'e-' + decimals);
/* eslint-enable prefer-template */

export const formatAmountToPrice = (amount, locale = DEFAULT_LOCALE) =>
  Number.parseFloat(amount).toLocaleString(locale, {
    minimumFractionDigits: MIN_FRACTION_DIGITS,
    maximumFractionDigits: MAX_FRACTION_DIGITS,
  });

/**
 * Curried format balance function to partially apply locale and then format numbers
 * @returns {Number}
 */
export const formatAmountWithLocale = curry((locale, balance) => formatAmountToPrice(balance, locale));
