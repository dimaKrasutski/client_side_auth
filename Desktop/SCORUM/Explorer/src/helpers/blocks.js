import { compose, last, dec, head, reverse } from 'ramda';
import { Subject } from 'rxjs/Subject';
import { getBlocksHistory } from '../api/scorum';

/**
 * Get first transaction from a list and take index property decrementing it
 */
export const getNextOffset = compose(dec, head, last);

/**
 * @param {function} reducer - function to reduce entity items
 * @param {number} blockNumber - block number
 * @param {function} setCurrentBlock - set current block
 */
export const createBlocksFetcher = (reducer, blockNumber, setCurrentBlock) => {
  const transactionsSubject = new Subject();
  let blockNum = blockNumber;
  let isStopped = false;
  let inProgress = false;

  transactionsSubject.hasNext = () => !isStopped;

  transactionsSubject.fetchNext = function fetchNext() {
    if (isStopped || inProgress) {
      return null;
    }

    const getBlocks = getBlocksHistory(blockNum);
    inProgress = true;
    getBlocks.subscribe(compose(
      (fetched) => {
        this.next(fetched);
        blockNum = getNextOffset(fetched);
        setCurrentBlock(blockNum);
        if (Number.isNaN(blockNum) || blockNum < 0) { // When we trap to negative - no more records left
          isStopped = true;
        }
        inProgress = false;
      },
      reverse,
    ));

    return null;
  };

  return transactionsSubject;
};
