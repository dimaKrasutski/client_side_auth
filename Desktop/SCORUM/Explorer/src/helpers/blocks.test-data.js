export const transactionHistorySourceWithZeroIndex = [
  [
    0,
    {
      block: 99,
      op: [
        'transfer_to_vesting',
        {
          amount: '1.700000000 SCR',
          from: 'songodzen',
          to: 'songodzen',
        },
      ],
      timestamp: '2018-03-07T15:42:39',
      trx_id: '9aaf9b03dcea97a79ccbb0771d6bc68c48cd7e67',
      trx_in_block: 5,
      virtual_op: 0,
    },
  ],
];

export const transactionHistorySource = [
  [
    7,
    {
      block: 100,
      op: [
        'transfer_to_vesting',
        {
          amount: '5.420000000 SCR',
          from: 'songodzen',
          to: 'songodzen',
        },
      ],
      timestamp: '2018-03-04T13:11:00',
      trx_id: '8d108dafb2144a1facd275f72858abc994ad0d76',
      trx_in_block: 5,
      virtual_op: 0,
    },
  ],
  [
    6,
    {
      block: 99,
      op: [
        'transfer_to_vesting',
        {
          amount: '1.700000000 SCR',
          from: 'songodzen',
          to: 'songodzen',
        },
      ],
      timestamp: '2018-03-07T15:42:39',
      trx_id: '9aaf9b03dcea97a79ccbb0771d6bc68c48cd7e67',
      trx_in_block: 5,
      virtual_op: 0,
    },
  ],
];
