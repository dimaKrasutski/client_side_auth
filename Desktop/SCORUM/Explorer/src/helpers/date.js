import dateformat from 'dateformat';
import { DEFAULT_LOCALE, LOCALE_KEYS } from '../constants';

/**
 * @TODO: Remove UTC when we decide to switch to format dates based on current user Timezone
 */
export const LOCALE_DATE_TIME_FORMATS = {
  [LOCALE_KEYS.EN_US]: 'UTC:dd/mm/yy h:MM TT',
  [LOCALE_KEYS.RU_RU]: 'UTC:dd/mm/yy HH:MM',
};

export const toStandardFormat = (timestamp, locale = DEFAULT_LOCALE) => {
  const localeDateTimeFormat = LOCALE_DATE_TIME_FORMATS[locale] || LOCALE_DATE_TIME_FORMATS[DEFAULT_LOCALE];
  // Convert timestamp passed to UTC format using 'Z' symbol at the end (if not defined already)
  const dateFromTimestamp = new Date(timestamp.replace(/(zZ)*$/, 'Z'));
  return dateformat(dateFromTimestamp, localeDateTimeFormat);
};
