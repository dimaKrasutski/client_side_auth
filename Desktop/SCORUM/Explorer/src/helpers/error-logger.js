
export const logError = (error) => {
  console.info('Error here!: ', error);
  return error;
};
