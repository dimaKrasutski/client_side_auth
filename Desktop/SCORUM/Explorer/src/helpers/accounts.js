import { compose, head, pick, curry, multiply, nAry } from 'ramda';
import { DEFAULT_LOCALE, MIN_FRACTION_DIGITS } from '../constants';
import { floorFloatAmount } from '../helpers/amount';

export const SCR_BALANCE_RESPONSE_FIELD_NAME = 'balance';
export const SP_BALANCE_RESPONSE_FIELD_NAME = 'scorumpower';


export const getAccountInfo = compose(
  pick([SCR_BALANCE_RESPONSE_FIELD_NAME, SP_BALANCE_RESPONSE_FIELD_NAME]),
  head,
);

export const balanceToNumber = nAry(1, Number.parseFloat);

export const formatBalanceToPrice = (balance, locale = DEFAULT_LOCALE) =>
  floorFloatAmount(balanceToNumber(balance), MIN_FRACTION_DIGITS).toLocaleString(locale, {
    minimumFractionDigits: MIN_FRACTION_DIGITS,
    maximumFractionDigits: MIN_FRACTION_DIGITS,
  });

/**
 * Curried format balance function to partially apply locale and then format numbers
 * @returns {Number}
 */
export const formatBalanceWithLocale = curry((locale, balance) => formatBalanceToPrice(balance, locale));

/**
 * Gets balance as string and returns balance as USD
 * @returns {Number}
 */
export const getBalanceConverter = rate => compose(
  multiply(rate),
  balanceToNumber,
);
