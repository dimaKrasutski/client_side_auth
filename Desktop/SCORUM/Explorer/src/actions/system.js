import { createActions } from 'redux-actions';
import { logError } from '../helpers/error-logger';

export const {
  globalError,
  clearError,
  redirectTo404,
  clearRedirect,
} = createActions(
  {
    GLOBAL_ERROR: (errorId, error, links) => {
      if (error) {
        logError(error);
      }
      return { error: errorId, links };
    },
  },
  'CLEAR_ERROR',
  'REDIRECT_TO_404',
  'CLEAR_REDIRECT',
);
