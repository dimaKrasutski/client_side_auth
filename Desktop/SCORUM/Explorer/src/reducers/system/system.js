import { handleActions } from 'redux-actions';
import { globalError, clearError, redirectTo404, clearRedirect } from '../../actions/system';
import { UNDEFINED_ERROR_ID } from '../../constants';
import initialState from './initial-state';

export const system = handleActions(
  {
    [globalError]: (state, { payload: { error = UNDEFINED_ERROR_ID, links } }) => ({
      ...state,
      globalError: {
        ...state.globalError,
        id: error,
        links,
      },
    }),
    [clearError]: state => ({
      ...state,
      globalError: initialState.globalError,
    }),
    [redirectTo404]: state => ({
      ...state,
      redirect404: true,
    }),
    [clearRedirect]: state => ({
      ...state,
      redirect404: false,
    }),
  },
  initialState,
);

