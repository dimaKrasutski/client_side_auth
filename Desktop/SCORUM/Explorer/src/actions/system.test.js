import { globalError } from './system';
import { logError } from '../helpers/error-logger';

jest.mock('../helpers/error-logger');

describe('system action creators', () => {
  test('should create global error with error id', () => {
    expect(globalError('test-id')).toMatchSnapshot();
    expect(logError).not.toHaveBeenCalled();
  });

  test('should create global error with error id and error', () => {
    const error = new Error('test');
    expect(globalError('test-id', error)).toMatchSnapshot();
    expect(logError).toHaveBeenCalledWith(error);
  });

  test('should create global error with error id, error and links', () => {
    const error = new Error('test');
    expect(globalError('test-id', error, [{ id: 'id', href: 'href', messageId: 'messageId' }])).toMatchSnapshot();
    expect(logError).toHaveBeenCalledWith(error);
  });
});
