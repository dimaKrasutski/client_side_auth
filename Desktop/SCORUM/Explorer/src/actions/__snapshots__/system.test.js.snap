// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`system action creators should create global error with error id 1`] = `
Object {
  "payload": Object {
    "error": "test-id",
    "links": undefined,
  },
  "type": "GLOBAL_ERROR",
}
`;

exports[`system action creators should create global error with error id and error 1`] = `
Object {
  "payload": Object {
    "error": "test-id",
    "links": undefined,
  },
  "type": "GLOBAL_ERROR",
}
`;

exports[`system action creators should create global error with error id, error and links 1`] = `
Object {
  "payload": Object {
    "error": "test-id",
    "links": Array [
      Object {
        "href": "href",
        "id": "id",
        "messageId": "messageId",
      },
    ],
  },
  "type": "GLOBAL_ERROR",
}
`;
